plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,237 @@
+                                   []      -> 1 -- depth of toplevel lambda
+                                   (d,_):_ -> d+1 -- increase depth
+                               , n^.PLC.unique.coerced) : ctx)
+
+
+floatTerm :: (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique, Monoid a)
+          => Term tyname name a -> Term tyname name a
+floatTerm pir =
+  let depthData = toDepthData $ compRanks pir -- run the first pass
+  in visitBody (topRank^._1) depthData (topRank^._2) pirClean
+ where
+  -- Visiting a term to apply the float transformation
+  visitTerm curDepth remInfo = \case
+      LamAbs a n ty tBody -> LamAbs a n ty $ visitBody  (curDepth+1) remInfo (n ^. PLC.unique . coerced) tBody
+      TyAbs a n k tBody  -> TyAbs a n k $ visitBody  (curDepth+1)  remInfo (n ^. PLC.unique . coerced) tBody
+      t -> over termSubterms (visitTerm curDepth remInfo) t -- descend
+
+  -- Special case of 'visitTerm' where we visit a lambda or a Lambda body
+  visitBody _  [] _ tBody = tBody -- no lets are left to float, do not descend
+  visitBody curDepth depthTable@((searchingForDepth, lams_lets):restDepthTable) u tBody
+    | curDepth < searchingForDepth = visitTerm curDepth depthTable tBody
+    | curDepth == searchingForDepth =
+        let tBody' = visitTerm curDepth restDepthTable tBody
+        in case M.lookup u lams_lets of
+             Just lets -> wrapLets lets curDepth restDepthTable tBody'
+             _         -> tBody'
+    | otherwise = error "just for completion"
+
+  -- TODO: we can transform easily the above visits to a Reader CurDepth
+  -- TODO: using the above pure/local way has the disadvantage that we visit a bit more than we need. To fix this we can use State DepthDataRemaining instead.
+
+
+  (pirClean, letTable) = removeLets pir
+
+  -- the classic dependency graph, same as the one used by the deadcode elimination transformation
+  depGraph = AM.gmap (\case Variable u -> u; _ -> error "just for completion")
+             $ AM.removeVertex Root -- we remove Root because we do not care about it right now
+             $ runTermDeps pir
+
+  -- the dependency graph, but with datatype-related nodes merged/grouped under a single node per datatypebind a, see 'bindingUnique'.
+  reducedDepGraph = M.foldr (\ (_,_,b) accGraph ->
+                              case b of
+                                DatatypeBind _ dt -> let princ = bindingUnique b
+                                                     in foldr (\ assocB -> AM.replaceVertex assocB princ) accGraph $ datatypeIds dt
+                                _ -> accGraph) depGraph letTable
+
+  -- take the strongly-connected components of the reduced dep graph, because it may contain loops (introduced by the LetRecs)
+  -- topologically sort these sccs, since we rely on linear (sorted) scoping in our 'wrapLets' code generation
+  topSortedSccs = fromJust $ AM.topSort $ AM.scc reducedDepGraph -- TODO: is this fromJust safe here?
```

Comment:
Never ever use `fromJust`, even if you're sure it's safe. Today it's safe, tomorrow someone changes something and it's no longer safe. The problem with `fromJust`, `head`, `tail` and other silly prelude functions is that they do not print a stack trace. They just output things like "head: empty list" and then go find where that error has happened.

Use `fromMaybe (error "<error_message>")` as `error` does print a stack trace.

Reply:
You are right, `fromJust` in general is not safe, but I thought maybe we could get away with this use of `fromJust`, since `topSort . scc` is guaranteed to return a `Just`? Anyway, I was also not sure about it, that is why I added a TODO.  I changed it to your suggestion and maybe later see how to hook up to the real compiler error reporting.